### PR TITLE
Clean up Storage/StorageImpl constructors

### PR DESCRIPTION
**Pull request by dzhulgakov**
Small cleanup while doing https://github.com/pytorch/pytorch/pull/16857:

- rename C2 constructors as create_legacy
- remove duplicated constructors
- make resizable flag non-default

### CODE REVIEW

**Review by smessmer — APPROVED** [review]

File: aten/src/ATen/templates/TypeDefault.cpp [hunk 1]
```diff
@@ -95,14 +95,17 @@ Tensor TypeDefault::tensorWithAllocator(IntArrayRef sizes, IntArrayRef strides,
...
       size,
-      deleter);
+      InefficientStdFunctionContext::makeDataPtr(
+          data, deleter, getDeviceFromPtr(data)),
+      /*allocator=*/nullptr,
+      /*resizable=*/false);
 }
 Storage TypeDefault::storageWithAllocator(int64_t size, Allocator* allocator) const {
-    return Storage(typeMeta(), size, allocator);
+  // Potentially the storage might be marked as resiable too here
```
Comment (gchanan) [inline aten/src/ATen/templates/TypeDefault.cpp]: nit: spelling.

File: c10/core/Storage.h [hunk 2]
```diff
@@ -53,6 +32,23 @@ struct C10_API Storage {
             allocator,
             resizable)) {}
 
+  static Storage create_legacy(at::DeviceType device_type) {
```
Comment (gchanan) [inline c10/core/Storage.h]: can we just kill this one?  It's only called from one place, and that place can just do this.

**Comment by ezyang** [comment]
This broke XLA build:

```
DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_XLAC -std=c++11
Feb 14 07:58:07 torch_xla/csrc/tensor_impl.cpp:115:10: error: no matching constructor for initialization of 'c10::Storage'
Feb 14 07:58:07   return c10::Storage(at::Device(c10::DeviceType::XLA, device.ordinal),
Feb 14 07:58:07          ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:10:3: note: candidate constructor not viable: requires single argument 'ptr', but 2 arguments were provided
Feb 14 07:58:07   Storage(c10::intrusive_ptr<StorageImpl> ptr) : storage_impl_(std::move(ptr)) {}
Feb 14 07:58:07   ^
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:7:16: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
Feb 14 07:58:07 struct C10_API Storage {
Feb 14 07:58:07                ^
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:7:16: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:9:3: note: candidate constructor not viable: requires 0 arguments, but 2 were provided
Feb 14 07:58:07   Storage() {}
Feb 14 07:58:07   ^
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:13:3: note: candidate constructor not viable: requires 4 arguments, but 2 were provided
Feb 14 07:58:07   Storage(
Feb 14 07:58:07   ^
Feb 14 07:58:07 /var/lib/jenkins/workspace/c10/core/Storage.h:27:3: note: candidate constructor not viable: requires 5 arguments, but 2 were provided
Feb 14 07:58:07   Storage(
Feb 14 07:58:07   ^
Feb 14 07:58:08 1 error generated.
```

cc @ailzhang

**Comment by ailzhang** [comment]
This should be fixed by https://github.com/pytorch/xla/pull/255, thanks @dlibenzi  @ezyang !

**Pull request closed without merging**